File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        StringBuilder str = new StringBuilder(s);
        int left = 0;
        int right = 0;
        int maxLen = 0;
        HashMap<Character, Integer> charMap = new HashMap<>();

        while( right < str.length()){
            if( charMap.containsKey(str.charAt(right))){
                left = Math.max(left, charMap.get(str.charAt(right)) + 1);
            }
            charMap.put( str.charAt(right), right);
            maxLen = Math.max(maxLen, right - left + 1);
            right++;
        }
        return maxLen;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    """  
    expand ---> if ferq[substring] <= 1
    shrink ---> if freq[substring] > 1

    """
    def lengthOfLongestSubstring(self, s: str) -> int:

        n = len(s)
        maxLen = 0
        freq = {}

        left = 0

        for right in range(n):
            freq[s[right]] = freq.get(s[right], 0) + 1

            while freq[s[right]] > 1:
                freq[s[left]] -= 1
                if freq[s[left]] == 0:
                    del freq[s[left]]
                left += 1
            maxLen = max( maxLen, right - left + 1)

        return maxLen